keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2384 @@
+ignore (or not try to manage itself). E.g. any non-namespaced QoS-class
+resources (one without `<namespace>/` prefix in the name) would be treated as
+Kubernetes-managed and ignored by the container runtime. See [Consts](#consts)
+section below for details about QoS-class resource naming.
+
+One possible usage-scenario would be pod-level cgroup controls, e.g. cgroup v2
+memory knobs in linux (see
+[KEP-2570: Support Memory QoS with cgroups v2][kep-2570].
+
+QoS-class resources could be used to specify OOM kill behavior of individual
+containers
+(ref [PR for disabling group oom kill](https://github.com/kubernetes/kubernetes/pull/122813)).
+
+Another possible usage could be to allow the user to explicitly specify the
+desired Pod QoS class of the application (instead of implicitly deriving it
+from the resource requests/limits).
+
+##### Container-level memory QoS
+
+Container runtimes could implement an admin-configurable support for memory
+QoS, as an alternative to
+[KEP-2570: Support Memory QoS with cgroups v2][kep-2570].
+
+It would be easy to enable also for example `memory.swap.*` control knobs
+available in Linux cgroups v2 to control swap-usage on a per-container basis.
+
+##### Runtime classes
+
+The QoS-class resources mechanism could be used to replace (or
+re-implement) the current runtime classes.
+
+One benefit of using a QoS-class resource to represent runtime classes would be
+the "automatic" visibility to what runtime classes are available on each node.
+E.g. adding a new runtime class on a node by re-configuration of the runtime
+would be immediately reflected in the node status without any additional admin
+tasks (like node labeling or manual creation of RuntimeClass objects).
+
+##### Splitting Pod QoS Class
+
+Currently the Pod QoS class is an implicit property of a Pod, tied to how the
+resource requests and limits of its containers are specified. QoS-class
+resources could make it possible for users to explicitly specify the Pod QoS
+class in the Pod's `.spec`, making it possible e.g. to define a pod with a
+container whose memory limit is higher than the requests, but that is treated
+by the node as `Guaranteed`.
+
+Taking this idea further, QoS-class resources could also make it possible to
+split several properties implicit in the Pod QoS class (for example: eviction
+behavior or OOM scoring) into separate properties. For example, in the case of
+OOM scoring make it possible for the user to specify a burstable pod with low
+OOM priority (low chance of being killed).
+
+##### Pod priority class
+
+One wild idea would be to implement pod priority class mechanism (or part of
+it) as a QoS-class resource. This would be a
+[Kubernetes-managed](#kubernetes-managed-qos-class-resources) pod-level
+QoS-class resource. Likely benefits of using the QoS-class resources mechanism
+would be to be able to set per-namespace defaults with LimitRanges and allow
+permission-control to high-priority classes with ResourceQuotas.
+
```

Comment:
Another idea is smarter oom setups (systemd-oomd for example).

systemd-oomd allows one to set settings for oom (psi based pressure and swap usage) on a cgroup slice. I could envision a future where one can create a QoS and that gives knobs to control systemd-oomd for more aggressive oom.

Reply:
https://fedoraproject.org/wiki/Changes/EnableSystemdOomd

I was playing around with this and I explored the ideas of setting different knobs based on the existing kubepods slices.

Reply:
Interesting. So you think a about pre-defined oom kill classes (in contrast to e.g. exposing every possible systemd-oomd knob to the user). At first thought, this sounds like a good fit. I'll add this to the proposal too, as a possible future use case.